File: comp/tools/screencast/tempo.mjs
// Screencast: the tempo worktime logger on the REAL running app. Seeds a few
// people, projects, and categories with time entries across the month via the
// API, then drives the SPA: a manager sees the whole team's distribution
// (donut by project, category + per-person bars, per-day series), flips the
// range (week/month/year) and the scope (Everyone/Mine), logs a live entry, and
// runs a pomodoro timer. Every chart is real report data from the component.
//
// Prereq: from repo root  `just host-tempo &`   (serves on :3040)
import { chromium } from "playwright";

const BASE = process.env.TEMPO_URL || "http://127.0.0.1:3040";
const OUT = new URL("./videos/tempo/", import.meta.url).pathname;
const W = 1120, H = 860;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// ---- seed via the API -------------------------------------------------------
const H_JSON = { "content-type": "application/json" };
async function api(path, method = "GET", body, token) {
  const h = { ...H_JSON }; if (token) h.authorization = `Bearer ${token}`;
  const r = await fetch(`${BASE}/api${path}`, { method, headers: h, body: body ? JSON.stringify(body) : undefined });
  return r.json().catch(() => ({}));
}
async function signup(email, role) {
  await api("/register", "POST", { email, password: "pw12345678", role });
  const l = await api("/login", "POST", { email, password: "pw12345678" });
  return l.access_token;
}
const month = new Date().toISOString().slice(0, 7); // current month, matches the SPA's default range
const day = (n) => `${month}-${String(n).padStart(2, "0")}`;

const admin = await signup("admin@acme.io", "admin");
await signup("boss@acme.io", "manager");
const boss = (await api("/login", "POST", { email: "boss@acme.io", password: "pw12345678" })).access_token;
const ada = await signup("ada@acme.io", "member");
const bo = await signup("bo@acme.io", "member");

// admin creates projects + categories
const P = {};
for (const [key, name] of [["APOLLO", "Apollo"], ["ZEPHYR", "Zephyr"], ["ORION", "Orion"]])
  P[name] = (await api("/projects", "POST", { key, name }, admin)).id;
const C = {};
for (const name of ["engineering", "sales", "design", "ops"])
  C[name] = (await api("/categories", "POST", { name }, admin)).id;

// spread entries across people, projects, categories, days
async function log(tok, proj, cat, d, mins) {
  await api("/entries", "POST", { project: P[proj], category: C[cat], minutes: mins, day: day(d) }, tok);
}
const plan = [
  [ada, "Apollo", "engineering", [[3,120],[5,180],[9,150],[12,90],[16,200],[19,160],[23,120]]],
  [ada, "Zephyr", "design", [[6,60],[13,90],[20,75]]],
  [bo, "Apollo", "engineering", [[4,140],[10,160],[17,130],[24,110]]],
  [bo, "Orion", "sales", [[2,90],[8,120],[15,140],[22,100]]],
  [boss, "Zephyr", "ops", [[7,80],[14,100],[21,60]]],
  [boss, "Orion", "engineering", [[11,90],[18,110]]],
];
for (const [tok, proj, cat, days] of plan) for (const [d, m] of days) await log(tok, proj, cat, d, m);

// ---- drive the SPA ----------------------------------------------------------
const browser = await chromium.launch({ headless: true });
const ctx = await browser.newContext({ viewport: { width: W, height: H }, recordVideo: { dir: OUT, size: { width: W, height: H } }, deviceScaleFactor: 2 });
const page = await ctx.newPage();
await page.goto(BASE);

try {
  // log in as the manager
  await page.fill("#email", "boss@acme.io");
  await page.fill("#password", "pw12345678");
  await page.click("#loginBtn");
  await page.locator("#app").waitFor({ timeout: 10000 });
  await sleep(1000);

  // the whole team's month: donut by project + category & per-person bars
  await page.click('#scopeSeg button[data-s="all"]');
  await sleep(2600);

  // flip the range
  await page.click('#ranges button[data-r="week"]'); await sleep(1600);
  await page.click('#ranges button[data-r="year"]'); await sleep(1600);
  await page.click('#ranges button[data-r="month"]'); await sleep(1400);

  // switch to my own view
  await page.click('#scopeSeg button[data-s="me"]'); await sleep(2000);

  // log a live entry — the charts update
  await page.selectOption("#proj", { label: "Orion" });
  await page.selectOption("#cat", { label: "engineering" });
  await page.fill("#mins", "45");
  await page.click("#logBtn");
  await sleep(2000);

  // run a pomodoro: start, let it tick, stop (logs an entry)
  await page.click("#pomo");
  await sleep(3200);
  await page.click("#pomo");
  await sleep(2200);
} finally {
  await ctx.close();
  await browser.close();
}
console.log("done");

File: comp/examples/tempo/tests/tempo.rs
//! E2E for the tempo worktime logger (TEMPO.md) as ONE composed wasm HTTP
//! component (tempo-domain + auth-guard + records) on the native Rust host.
//! Proves: auth + RBAC (admin creates projects/categories, member can't), time
//! logging + a pomodoro timer, and role-scoped range aggregation (a member sees
//! only their own totals; a manager sees the whole org, grouped by user).

use std::path::PathBuf;
use std::process::{Child, Command};
use std::time::Duration;

use serde_json::{json, Value};

const ADDR: &str = "127.0.0.1:3040";

struct HostGuard(Child);
impl Drop for HostGuard {
    fn drop(&mut self) {
        let _ = self.0.kill();
        let _ = self.0.wait();
    }
}

fn base() -> String {
    format!("http://{ADDR}")
}

fn req(method: &str, path: &str, token: Option<&str>, body: Option<Value>) -> (u16, Value) {
    let url = format!("{}{}", base(), path);
    let mut r = ureq::request(method, &url);
    if let Some(t) = token {
        r = r.set("authorization", &format!("Bearer {t}"));
    }
    let result = match &body {
        Some(b) => r.set("content-type", "application/json").send_string(&b.to_string()),
        None => r.call(),
    };
    let resp = match result {
        Ok(resp) => resp,
        Err(ureq::Error::Status(_, resp)) => resp,
        Err(e) => panic!("{method} {path}: {e}"),
    };
    let status = resp.status();
    (status, serde_json::from_str(&resp.into_string().unwrap_or_default()).unwrap_or(Value::Null))
}

fn signup(email: &str, role: &str) -> String {
    let (s, _) = req("POST", "/api/register", None, Some(json!({ "email": email, "password": "pw12345678", "role": role })));
    assert!(s == 201 || s == 409, "register {email}: {s}");
    let (s, l) = req("POST", "/api/login", None, Some(json!({ "email": email, "password": "pw12345678" })));
    assert_eq!(s, 200, "login {email}: {l}");
    l["access_token"].as_str().unwrap().to_string()
}

fn start_host() -> HostGuard {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..").canonicalize().unwrap();
    let bin = root.join("host/target/release/vet-host");
    let component = root.join("components/target/tempo_domain.composed.wasm");
    assert!(bin.exists(), "host not built: {bin:?} (run `just e2e-tempo`)");
    assert!(component.exists(), "composed wasm missing (just compose-tempo)");
    let child = Command::new(&bin)
        .args(["--component", component.to_str().unwrap(), "--addr", ADDR, "--kv", "memory"])
        .env("VET_TENANT", "tempo")
        .spawn()
        .expect("spawn vet-host");
    let guard = HostGuard(child);
    for _ in 0..200 {
        if let Ok(r) = ureq::get(&base()).call() {
            if r.status() == 200 {
                return guard;
            }
        }
        std::thread::sleep(Duration::from_millis(50));
    }
    panic!("tempo host did not start");
}

fn total(v: &Value) -> u64 {
    v["total_minutes"].as_u64().unwrap()
}

#[test]
fn worktime_logging_rbac_and_reports() {
    let _host = start_host();

    let admin = signup("admin@acme.io", "admin");
    let ada = signup("ada@acme.io", "member");
    let boss = signup("boss@acme.io", "manager");

    // ===== admin creates projects + categories; a member can't ==============
    let (s, proj) = req("POST", "/api/projects", Some(&admin), Some(json!({ "key": "APOLLO", "name": "Apollo" })));
    assert_eq!(s, 201, "create project: {proj}");
    let pid = proj["id"].as_str().unwrap().to_string();
    let (s, eng) = req("POST", "/api/categories", Some(&admin), Some(json!({ "name": "engineering" })));
    assert_eq!(s, 201, "{eng}");
    let eng_id = eng["id"].as_str().unwrap().to_string();
    let (_, sales) = req("POST", "/api/categories", Some(&admin), Some(json!({ "name": "sales" })));
    let sales_id = sales["id"].as_str().unwrap().to_string();

    let (s, _) = req("POST", "/api/projects", Some(&ada), Some(json!({ "key": "X", "name": "X" })));
    assert_eq!(s, 403, "a member cannot create projects");

    // ===== a member logs time ===============================================
    let log = |cat: &str, minutes: u64, day: &str| {
        let (s, e) = req("POST", "/api/entries", Some(&ada),
            Some(json!({ "project": pid, "category": cat, "minutes": minutes, "day": day })));
        assert_eq!(s, 201, "log: {e}");
    };
    log(&eng_id, 120, "2026-07-20");
    log(&sales_id, 60, "2026-07-21");
    // an unknown project is rejected
    let (s, _) = req("POST", "/api/entries", Some(&ada),
        Some(json!({ "project": "nope", "category": eng_id, "minutes": 30, "day": "2026-07-20" })));
    assert_eq!(s, 422, "unknown project rejected");

    // ===== the member's own report ==========================================
    let (s, r) = req("GET", "/api/report?from=2026-07-01&to=2026-07-31", Some(&ada), None);
    assert_eq!(s, 200, "{r}");
    assert_eq!(total(&r), 180, "member total: {r}");
    let by_proj = r["by_project"].as_array().unwrap();
    assert_eq!(by_proj.len(), 1);
    assert_eq!(by_proj[0]["name"], "Apollo");
    assert_eq!(by_proj[0]["minutes"], 180);
    let cats: Vec<(&str, u64)> = r["by_category"].as_array().unwrap().iter()
        .map(|c| (c["key"].as_str().unwrap(), c["minutes"].as_u64().unwrap())).collect();
    assert!(cats.contains(&("engineering", 120)) && cats.contains(&("sales", 60)), "by category: {cats:?}");

    // a member can't escalate scope — 'all' is forced back to 'me'
    let (_, r) = req("GET", "/api/report?from=2026-07-01&to=2026-07-31&scope=all", Some(&ada), None);
    assert_eq!(r["scope"], "me", "member scope stays me");
    assert_eq!(r["can_see_all"], false);

    // range filtering excludes out-of-range days
    let (_, r) = req("GET", "/api/report?from=2026-07-21&to=2026-07-31", Some(&ada), None);
    assert_eq!(total(&r), 60, "only the 21st: {r}");

    // ===== a manager sees the whole org, grouped by user ====================
    let (s, r) = req("GET", "/api/report?from=2026-07-01&to=2026-07-31&scope=all", Some(&boss), None);
    assert_eq!(s, 200, "{r}");
    assert_eq!(r["scope"], "all");
    assert_eq!(r["can_see_all"], true);
    assert_eq!(total(&r), 180, "manager sees the member's time: {r}");
    let users: Vec<&str> = r["by_user"].as_array().unwrap().iter().map(|u| u["key"].as_str().unwrap()).collect();
    assert!(users.contains(&"ada@acme.io"), "by_user has the member: {users:?}");

    // ===== the pomodoro timer ===============================================
    let (s, _) = req("POST", "/api/timer/start", Some(&ada),
        Some(json!({ "project": pid, "category": eng_id, "day": "2026-07-22" })));
    assert_eq!(s, 200);
    let (_, t) = req("GET", "/api/timer", Some(&ada), None);
    assert!(t["timer"]["started"].as_u64().is_some(), "timer running: {t}");
    let (s, entry) = req("POST", "/api/timer/stop", Some(&ada), None);
    assert_eq!(s, 201, "stop -> entry: {entry}");
    assert!(entry["minutes"].as_u64().unwrap() >= 1, "at least a minute");

    let (_, r) = req("GET", "/api/report?from=2026-07-01&to=2026-07-31", Some(&ada), None);
    assert!(total(&r) >= 181, "timer added to the total: {r}");
}
